instrumentor/controllers/instrumentationdevice/collectorsgroup_controller.go

```diff
@@ -18,34 +18,116 @@ package instrumentationdevice
 
 import (
 	"context"
+	"errors"
 
 	odigosv1 "github.com/odigos-io/odigos/api/odigos/v1alpha1"
 
+	"github.com/odigos-io/odigos/k8sutils/pkg/consts"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
 )
 
+// CollectorsGroupReconciler is responsible for reconciling the instrumented workloads
+// once the collectors group becomes ready - by adding the instrumentation device to the workloads.
+// This is necessary to ensure that we won't instrument any workload before the
+// node collectors are ready to receive the data.
 type CollectorsGroupReconciler struct {
 	client.Client
 	Scheme *runtime.Scheme
 }
 
-func (r *CollectorsGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+type collectorsGroupReadyPredicate struct {
```

Comment:
notice that in https://github.com/odigos-io/odigos/pull/1714 I extracted this predicate into `k8sutils` so you can refactor this part if we merge it first

Reply:
Done